### PR TITLE
add test script for rhel 8.4 ng virt test

### PR DESCRIPTION
**Pull request by yih-redhat**
Add ostree-ng.sh test in downstream to test rhel for edge on rhel 8.4 virt.

### CODE REVIEW

File: kitebot/Jenkinsfile [hunk 1]
```diff
@@ -175,6 +205,18 @@ void run_virt_test() {
     """
 }
 
+void run_84_virt_test() {
```
Comment (henrywang) [inline kitebot/Jenkinsfile:208]: Let's re-use `run_virt_test()` with arg added, things like `run_virt_test(test_type)`.
```
if (test_type == 'ostree-ng') {
    sh """
          git clone -b rhel-8.4.0 https://github.com/osbuild/osbuild-composer.git
         sudo mkdir -p /usr/libexec/osbuild-composer-test && sudo cp scripts/provision.sh /usr/libexec/osbuild-composer-test/provision.sh
          osbuild-composer/test/cases/ostree-ng.sh
     """
}
if (test_type == 'ostree-legacy') {
    sh "./rhel-edge-virt-test.sh "
}
```

File: rhel-edge-84-virt-test.sh [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+set -uo pipefail
+source /etc/os-release
+
+function greenprint {
+    echo -e "\033[1;32m${1}\033[0m"
+}
+
+# Initialize repo
+greenprint "Configure yum repository"
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:13]: We do not need configure repo here, it's already in jslave vm already.

File: rhel-edge-84-virt-test.sh [hunk 3]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+# Test script to test RHEL for edge on RHEL 8.4 virtual.
+# Will prepare the env, download upstream test script and run it.
+
+set -uo pipefail
+source /etc/os-release
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:6]: Does this script need this?

File: rhel-edge-84-virt-test.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+gpgcheck=0
+EOF
+
+greenprint "Install packages: git wget ansible httpd osbuild-composer composer-cli"
+sudo dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
+sudo dnf install -y --nogpgcheck osbuild-composer composer-cli git curl jq expect wget ansible podman httpd qemu-img qemu-kvm virt-install libvirt-client libvirt-daemon-kvm bash-completion
+sudo rpm -qa|grep osbuild
+
+greenprint "Start image builder service"
+sudo systemctl enable --now osbuild-composer.socket
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:33]: this starts osbuild-composer.socket and .service already. So the following two lines of codes are not required.
And you have a lot of osbuild-composer settings changed later, move this to the end of this file.

File: rhel-edge-84-virt-test.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+sudo systemctl start osbuild-composer.socket
+sudo systemctl start osbuild-composer.service
+
+greenprint "Git clone upstream project osbuild/osbuild-composer"
+sudo git clone https://github.com/osbuild/osbuild-composer.git
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:38]: 1. sudo is not required here.
2. please clone branch directly with `git clone -b rhel-8.4.0 https://github.com/osbuild/osbuild-composer.git`

File: rhel-edge-84-virt-test.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# shellcheck disable=SC2164
+cd osbuild-composer/
+sudo git checkout rhel-8.4.0
+# shellcheck disable=SC2103
+cd ..
+
+greenprint "Prepare osbuild-composer repository file"
+if [ ! -d /etc/osbuild-composer/repositories ]; then
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:46]: mkdir -p will not fail if folder exists already, so `-d` can be removed.

File: rhel-edge-84-virt-test.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    sudo mkdir -p /etc/osbuild-composer/repositories
+fi
+sudo cp osbuild-composer/test/data/repositories/rhel-84.json /etc/osbuild-composer/repositories/rhel-8-beta.json
+
+greenprint "Prepare RSA key files"
+if [ ! -d /usr/share/tests/osbuild-composer/keyring ]; then
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:52]: same as above.

File: rhel-edge-84-virt-test.sh [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    sudo mkdir -p /usr/share/tests/osbuild-composer/keyring
+fi
+sudo cp osbuild-composer/test/data/keyring/id_rsa /usr/share/tests/osbuild-composer/keyring/
+sudo chmod 600 /usr/share/tests/osbuild-composer/keyring/*
+
+# Copy provision file to target place, which will be called by ostree-ng.sh
+greenprint "Prepare provision file"
+if [ ! -d /usr/libexec/osbuild-composer-test ]; then
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:60]: same as above

File: rhel-edge-84-virt-test.sh [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    sudo mkdir -p /usr/libexec/osbuild-composer-test
+    sudo cp ./scripts/provision.sh /usr/libexec/osbuild-composer-test/provision.sh
+fi
+sudo cp ./scripts/provision.sh /usr/libexec/osbuild-composer-test/provision.sh
+
+greenprint "Prepare ansible playbook"
+if [ ! -d /usr/share/tests/osbuild-composer/ansible ]; then
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:67]: same as above

File: rhel-edge-84-virt-test.sh [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    sudo mkdir -p /usr/share/tests/osbuild-composer/ansible
+fi
+sudo cp osbuild-composer/test/data/ansible/check_ostree.yaml /usr/share/tests/osbuild-composer/ansible/
+
+sleep 10
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:72]: Why sleep here?

File: rhel-edge-84-virt-test.sh [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+sudo systemctl start osbuild-composer.service
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:73]: Move line 33 here.

File: rhel-edge-84-virt-test.sh [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+WORKSPACE=$(pwd)
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:75]: please do not defnine `WORKSPACE` here. It's pre-defined by Jenkins.

File: scripts/provision.sh [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Comment (henrywang) [inline scripts/provision.sh:1]: Move all content in `rhel-edge-84-virt-test.sh` into this file, and do not need `rhel-edge-84-virt-test.sh`.

**Review by henrywang — CHANGES_REQUESTED** [review]
Nice! Thanks @yih-redhat. You have my comments.
BTW: PSI OpenStack network will be updated from OVS to OVN. All tests will not be run in this case. Let's re-run it on Monday.

**Review by henrywang — CHANGES_REQUESTED** [review]

File: rhel-edge-84-virt-test.sh (hunk 2 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:13]: ok. will remove it.

File: rhel-edge-84-virt-test.sh (hunk 3 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:6]: ostree-ng.sh will do the same thing, will remove it.

File: rhel-edge-84-virt-test.sh (hunk 4 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:33]: ok.

File: rhel-edge-84-virt-test.sh (hunk 6 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:46]: ok. got it.

File: kitebot/Jenkinsfile (hunk 1 above)
Comment (yih-redhat) [reply kitebot/Jenkinsfile:208]: ok, let use only one entry script.

File: rhel-edge-84-virt-test.sh (hunk 7 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:52]: ok

File: rhel-edge-84-virt-test.sh (hunk 8 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:60]: ok

File: rhel-edge-84-virt-test.sh (hunk 9 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:67]: ok

File: rhel-edge-84-virt-test.sh (hunk 10 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:72]: it's a workaround in my env, sometimes osbuild service is not started after I run this command:
sudo systemctl enable --now osbuild-composer.socket

I don't know why and cannot find the root reason, so I wait a while and restart again to make sure this service is started.

I will check it again when I update the script according to your comments.

File: rhel-edge-84-virt-test.sh (hunk 11 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:73]: ok

File: rhel-edge-84-virt-test.sh (hunk 12 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:75]: ok

File: rhel-edge-84-virt-test.sh [hunk 14]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+# Test script to test RHEL for edge on RHEL 8.4 virtual.
+# Will prepare the env, download upstream test script and run it.
+
+set -uo pipefail
```
[the remark this answers is not in the document]
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:5]: ok

File: rhel-edge-84-virt-test.sh [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+source /etc/os-release
+
+function greenprint {
```
[the remark this answers is not in the document]
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:8]: ok

File: scripts/provision.sh (hunk 13 above)
Comment (yih-redhat) [reply scripts/provision.sh:1]: rhel-edge-84-virt-test.sh will checkout git branch
provision.sh is a file that required by the ostree-ng.sh in the git branch.

so, you mean we only do checkout in rhel-edge-84-virt-test.sh, and move all other setup script to provision.sh?

File: rhel-edge-84-virt-test.sh [hunk 16]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (henrywang) [inline rhel-edge-84-virt-test.sh:1]: Why do you still keep this file? All things are duplicated with what you added in Jenkinsfile

File: scripts/provision.sh [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+# This file is here because upstream test scripts ostree-ng.sh will call it.
+#
+# The steps in this file are moved to rhel-edge-84-virt-test.sh, which is the main script to test RHEL 8.4
+set -exuo pipefail
+
+echo "===>Install packages"
```
Comment (henrywang) [inline scripts/provision.sh:7]: You add -x in this script, so `echo` is not necessary in this case.

File: scripts/provision.sh [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+sudo mkdir -p /usr/libexec/osbuild-composer-test
+sudo cp ./scripts/provision.sh /usr/libexec/osbuild-composer-test/provision.sh
+
+echo "===>Prepare ansible playbook"
+sudo mkdir -p /usr/share/tests/osbuild-composer/ansible
+sudo cp osbuild-composer/test/data/ansible/check_ostree.yaml /usr/share/tests/osbuild-composer/ansible/
+
+echo "===>Start image builder service"
+sudo systemctl enable --now osbuild-composer.socket
+#sudo systemctl start osbuild-composer.socket
```
Comment (henrywang) [inline scripts/provision.sh:32]: Please remove comments.

**Review by henrywang — CHANGES_REQUESTED** [review]

File: scripts/provision.sh [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+echo "===>Prepare osbuild-composer repository file"
+sudo mkdir -p /etc/osbuild-composer/repositories
+sudo cp osbuild-composer/test/data/repositories/rhel-84.json /etc/osbuild-composer/repositories/rhel-8-beta.json
+
+echo "===>Prepare RSA key files"
+sudo mkdir -p /usr/share/tests/osbuild-composer/keyring
+sudo cp osbuild-composer/test/data/keyring/id_rsa /usr/share/tests/osbuild-composer/keyring/
+sudo chmod 600 /usr/share/tests/osbuild-composer/keyring/*
+
+# Copy provision file to target place, which will be called by ostree-ng.sh
```
Comment (henrywang) [inline scripts/provision.sh:21]: This's duplicated with Jenkinsfile's

File: scripts/provision.sh [hunk 20]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+# This file is here because upstream test scripts ostree-ng.sh will call it.
+#
+# The steps in this file are moved to rhel-edge-84-virt-test.sh, which is the main script to test RHEL 8.4
```
Comment (henrywang) [inline scripts/provision.sh:4]: This comment is not correct after your update.

**Review by henrywang — CHANGES_REQUESTED** [review]

**Comment by henrywang** [comment]
Could you please add the following code at the end of `provision.sh`? We need to make sure everything is ready to go. Thanks.
```
# Basic verification
sudo composer-cli status show
sudo composer-cli sources list
for SOURCE in $(sudo composer-cli sources list); do
    sudo composer-cli sources info "$SOURCE"
done
```

File: scripts/provision.sh (hunk 20 above)
Comment (yih-redhat) [reply scripts/provision.sh:4]: ok, will fix it.

File: scripts/provision.sh (hunk 17 above)
Comment (yih-redhat) [reply scripts/provision.sh:7]: got it.

File: scripts/provision.sh (hunk 18 above)
Comment (yih-redhat) [reply scripts/provision.sh:32]: ok

File: rhel-edge-84-virt-test.sh (hunk 16 above)
Comment (yih-redhat) [reply rhel-edge-84-virt-test.sh:1]: I didn't realize that my commit will reflect in this pull request.

I run test in a vm, and I usually pushed to git and then download in vm to run test, this file is just for test.

This commit should not be in this pull request, it's just for test.

Anyway, I will close this pull request as there is too many conversations here, and create a new clean on.

File: scripts/provision.sh (hunk 19 above)
Comment (yih-redhat) [reply scripts/provision.sh:21]: got it.

**Comment by yih-redhat** [comment]
Got it, I will start a new pull request, as this one has too many conversations.

Will add this to provision.sh

# Basic verification
sudo composer-cli status show
sudo composer-cli sources list
for SOURCE in $(sudo composer-cli sources list); do
    sudo composer-cli sources info "$SOURCE"
done

**Pull request closed without merging**